### PR TITLE
Reinstate updated python parser

### PR DESCRIPTION
**Pull request by littleforest**
#### :zap: Summary

This reinstates the work done in #116 that was rolled back in #120. We found that users sometimes had a `.coverage` file, but did not have the `coverage` program installed, causing an error to be raised.

This adds a check for the `coverage` executable in the `matches?` method for the python parser. If a file matches the `python` format, then we also do a secondary check to make sure that `coverage` is installed. If not, then we output a message about the missing program when the `debug` flag is passed.

### CODE REVIEW

File: src/coverage_reporter/parsers/coveragepy_parser.cr [hunk 1]
```diff
@@ -21,123 +18,54 @@ module CoverageReporter
...
 
     def matches?(filename : String) : Bool
-      File.open(filename) do |f|
+      valid_file_exists = File.open(filename) do |f|
         f.read_at(0, 15) do |io|
           io.gets.try(&.downcase) == "sqlite format 3"
         end
       end
+
+      check_for_coverage_executable(valid_file_exists)
```
Comment (mike-burns) [inline src/coverage_reporter/parsers/coveragepy_parser.cr:27]: IS there a reason to pass the Boolean instead of e.g.

```suggestion
      valid_file_exists && check_for_coverage_executable
```

**Review by mike-burns — APPROVED** [review]
I'm optimistic about this solution.

File: src/coverage_reporter/parsers/coveragepy_parser.cr (hunk 1 above)
Comment (littleforest) [reply src/coverage_reporter/parsers/coveragepy_parser.cr]: Oh yes, that's way better, thanks! Fixed in 0339f0321c6afb11d224d97c79b08612c51185d0

**Review by afinetooth — CHANGES_REQUESTED** [review]
**tl;dr**:
Requested changes:
- Can we add more copy to the error message (content suggestion below).
- Can we / shouldn't we also deliver the error even when `--debug` flag is not used? Otherwise, how will the user know why the upload failed? (Correct me if I'm wrong, but we're just going to bubble up whatever error we got from the environment, right?)

**Details**:
I think it will work for most cases. However, I am concerned about what happens when a user receives the error about the `coverage` executable not being present and them objecting to making that present in the environment (perhaps it's a challenge or requires a workaround they don't want to consider). In that case, there is no way to get from the `.coverage` file to the `coverage.xml` file, correct?

Given that, I thin the minimum change here is adding to the error copy stating that the user can solve their problem in one of two ways: (A) Add/keep the `coverage` executable in their (runner) environment; or (B) Convert their `.coverage` file into `coverage.xml` beforehand with `coverage xml` and change their input option to `format: cobertura` (for Coveralls action or orb), or `--format=cobertura` (for coverage reporter alone).

**Comment by littleforest** [comment]
> Can we / shouldn't we also deliver the error even when --debug flag is not used? Otherwise, how will the user know why the upload failed?

We can deliver it all the time. However, the upload might not have failed in the case that the user is not specifying a format. This was trying to cover the situation where the user has a `coverage.xml` file (for example), but also a `.coverage` file. The `coverage.xml` file goes through the Cobertura parser just fine, but then the `.coverage` file gets hung up because the user didn't have `coverage` installed. The changes here are such that we don't consider the `.coverage` file to "match" the requirements for the python parser if `coverage` is not installed, and we just note that in the output message.

**Comment by afinetooth** [comment]
I left some additional comments in our Slack thread on this, [here](https://coveralls.slack.com/archives/C060M754XSR/p1713299890836329?thread_ts=1713293409.196559&cid=C060M754XSR).

> We can deliver it all the time. However, the upload might not have failed in the case that the user is not specifying a format. This was trying to cover the situation where the user has a `coverage.xml` file (for example), but also a `.coverage` file. The `coverage.xml` file goes through the Cobertura parser just fine, but then the `.coverage` file gets hung up because the user didn't have `coverage` installed. 

OK, got it. In that case, I would say the determining factor as to whether to deliver an error is:

1. If the user specified `format: cobertura` - Don't deliver an error. The integration had what it needed to parse and upload a coverage report.
2. If the user specified `format: python` - Deliver an error in both `--debug` and non-`--debug` states. The integration cannot meet the requirement of uploading a `.coverage` report (without `coverage` being present).
3. If the user did not specify format (`format: null`) - Only deliver an "error" / "warning" message in `--debug` state. We don't need to deliver the error message outside of `--debug` state because the integration can meet the requirement of uploading _some_ coverage report. but it would be helpful to let the user know that there was a coverage report (in `python` format) that the integration found but could not upload (via its `python` parser).

WDYT?

> The changes here are such that we don't consider the `.coverage` file to "match" the requirements for the python parser if `coverage` is not installed, and we just note that in the output message.

Agreed.

I just think that in each of the states above it the message a different level of import. In case (2) it's an **error**. In case (3) it's a **warning**, because at least one coverage report was uploaded.

**Comment by littleforest** [comment]
@afinetooth here are the following updates

> 1. If the user specified format: cobertura - Don't deliver an error. The integration had what it needed to parse and upload a coverage report.

This will not hit the python parser at all (existing behavior), so no warning or error will be displayed.

> 2. If the user specified format: python - Deliver an error in both --debug and non---debug states. The integration cannot meet the requirement of uploading a .coverage report (without coverage being present).

When a format is specified, the code does not use the `matches?` method so the check in the `matches?` method to see if `coverage` is installed will be bypassed. Instead, the code will attempt to use the `coverage` executable on the file directly. Currently this will raise an error if `coverage` is not installed, but me know if you would prefer to use `Log.error` (see 20f872bc2442926c1ec82cafe5aef39a7e4cf196 for the copy output).

> 3. If the user did not specify format (format: null) - Only deliver an "error" / "warning" message in --debug state. 

This will deliver a warning in `--debug` state: 

```
☝️ Detected coverage format: python, but error with coverage executable: sh: coverage: command not found
```

**Comment by afinetooth** [comment]
@littleforest this looks good, thank you! I think these choices are fine and reasonable. If feedback differs we can iterate from here. Thanks again.

**Review by afinetooth — APPROVED** [review]
Looks good. Thank you!
